kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClient.java

```diff
@@ -0,0 +1,128 @@
+import io.netty.channel.ChannelOption;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+
+import io.kroxylicious.test.codec.DecodedRequestFrame;
+import io.kroxylicious.test.codec.DecodedResponseFrame;
+import io.kroxylicious.test.codec.KafkaApiMessageConverter;
+import io.kroxylicious.test.codec.KafkaRequestEncoder;
+import io.kroxylicious.test.codec.KafkaResponseDecoder;
+import io.kroxylicious.test.yaml.YamlMapper;
+
+import static io.kroxylicious.test.codec.KafkaApiMessageConverter.requestConverterFor;
+
+/**
+ * KafkaClient for testing.
+ * <p>
+ * The kafka client closes its channel after it gets a response. A new
+ * connection is bootstrapped on every call to `get`.
+ * </p>
+ * <p>
+ * The intention is that it should be used to fire a single RPC at a server
+ * offering the Kafka protocol, read a response and inform the client of that
+ * response. It currently translates the response to a normalised JsonObject.
+ * </p>
+ */
+public final class KafkaClient implements AutoCloseable {
+
+    /**
+     * create empty kafkaClient
+     */
+    public KafkaClient() {
+    }
+
+    private final EventLoopGroup group = new NioEventLoopGroup();
+
+    private static final AtomicInteger correlationId = new AtomicInteger(1);
+
+    private static DecodedRequestFrame<?> toApiRequest(JsonNode jsonNode) {
+        var apiMessageTypeNode = jsonNode.get("apiMessageType");
+        var versionNode = jsonNode.get("version");
+        var messageType = ApiMessageType.valueOf(apiMessageTypeNode.asText());
+        var header = new RequestHeaderData().setRequestApiKey(messageType.apiKey()).setRequestApiVersion(versionNode.shortValue());
+        header.setClientId(jsonNode.has("clientId") ? jsonNode.get("clientId").asText() : null);
+        header.setCorrelationId(correlationId.incrementAndGet());
+        var requestConverter = requestConverterFor(messageType);
+        JsonNode requestNode = jsonNode.get("request");
+        var request = requestConverter.reader().apply(requestNode, header.requestApiVersion());
+        return new DecodedRequestFrame<>(header.requestApiVersion(), header.correlationId(), header, request);
+    }
+
+    // TODO return a Response class with jsonObject() and frame() methods
+
+    /**
+     * Bootstrap and connect to a Kafka broker on a given host and port. Send
+     * the request to it and inform the client when we have received a response.
+     * The channel is closed after we have received the message.
+     * @param host kafka broker host
+     * @param port kafka broker port
+     * @param request request to send to kafka
+     * @return a future that will be completed with the response from the kafka broker (translated to JsonNode)
+     */
+    public CompletableFuture<JsonNode> get(String host, int port, JsonNode request) {
```

Comment:
I think we should move away from JsonNode to a `Response` class, maybe with a method like `toJsonNode()` on it to enable YAML/json driven declarative tests later.

Reply:
Done, the mocks/client no longer operate on jackson JsonNode